File: services/media-transform-runtime-service/src/handlers/image_ops.rs
//! Native image transformations backed by the [`image`] crate.
//!
//! Every handler here:
//!   * Decodes via `image::load_from_memory_with_format` so the input
//!     MIME drives the codec choice (no auto-sniffing — the REST layer
//!     already knows the source MIME).
//!   * Encodes back to the same format unless the params explicitly
//!     ask for a different one (`output_format`).
//!
//! Output bytes are returned as `Vec<u8>`; the REST layer base64-encodes
//! them at the boundary.

use image::{DynamicImage, ImageFormat, imageops::FilterType};
use serde::Deserialize;

use super::{HandlerError, HandlerOutput, HandlerResult};

fn format_for_mime(mime: &str, kind: &str) -> Result<ImageFormat, HandlerError> {
    Ok(match mime {
        "image/png" => ImageFormat::Png,
        "image/jpeg" | "image/jpg" => ImageFormat::Jpeg,
        "image/webp" => ImageFormat::WebP,
        "image/gif" => ImageFormat::Gif,
        "image/tiff" => ImageFormat::Tiff,
        "image/bmp" => ImageFormat::Bmp,
        other => {
            return Err(HandlerError::UnsupportedMime(
                other.to_string(),
                kind.to_string(),
            ));
        }
    })
}

fn load(bytes: &[u8], mime: &str, kind: &str) -> Result<DynamicImage, HandlerError> {
    let format = format_for_mime(mime, kind)?;
    image::load_from_memory_with_format(bytes, format)
        .map_err(|err| HandlerError::Decode(err.to_string()))
}

fn encode(img: DynamicImage, format: ImageFormat) -> Result<Vec<u8>, HandlerError> {
    let mut buf = std::io::Cursor::new(Vec::new());
    img.write_to(&mut buf, format)
        .map_err(|err| HandlerError::Encode(err.to_string()))?;
    Ok(buf.into_inner())
}

#[derive(Debug, Deserialize, Default)]
struct ResizeParams {
    width: Option<u32>,
    height: Option<u32>,
    /// Default `Lanczos3` — Foundry's "Resize" produces high-fidelity
    /// downscales; matches the upstream sample renders.
    #[serde(default)]
    filter: Option<String>,
}

fn parse_filter(name: Option<&str>) -> FilterType {
    match name.unwrap_or("lanczos3") {
        "nearest" => FilterType::Nearest,
        "triangle" => FilterType::Triangle,
        "catmullrom" => FilterType::CatmullRom,
        "gaussian" => FilterType::Gaussian,
        _ => FilterType::Lanczos3,
    }
}

pub fn thumbnail(mime: &str, params: &serde_json::Value, bytes: Vec<u8>) -> HandlerResult {
    // Foundry's thumbnail is a fixed 256-px longest-edge box.
    let img = load(&bytes, mime, "thumbnail")?;
    let max_dim = params.get("max_dim").and_then(|v| v.as_u64()).unwrap_or(256) as u32;
    let scaled = img.thumbnail(max_dim, max_dim);
    let format = format_for_mime(mime, "thumbnail")?;
    Ok(HandlerOutput {
        output_mime_type: mime.to_string(),
        output_bytes: Some(encode(scaled, format)?),
        output_json: None,
    })
}

pub fn resize(mime: &str, params: &serde_json::Value, bytes: Vec<u8>) -> HandlerResult {
    let p: ResizeParams = serde_json::from_value(params.clone())
        .map_err(|err| HandlerError::InvalidParams("resize".into(), err.to_string()))?;
    let img = load(&bytes, mime, "resize")?;
    let target_w = p.width.unwrap_or(img.width());
    let target_h = p.height.unwrap_or(img.height());
    let resized = img.resize_exact(target_w, target_h, parse_filter(p.filter.as_deref()));
    let format = format_for_mime(mime, "resize")?;
    Ok(HandlerOutput {
        output_mime_type: mime.to_string(),
        output_bytes: Some(encode(resized, format)?),
        output_json: None,
    })
}

pub fn resize_within_bbox(
    mime: &str,
    params: &serde_json::Value,
    bytes: Vec<u8>,
) -> HandlerResult {
    let p: ResizeParams = serde_json::from_value(params.clone()).map_err(|err| {
        HandlerError::InvalidParams("resize_within_bounding_box".into(), err.to_string())
    })?;
    let max_w = p.width.ok_or_else(|| {
        HandlerError::InvalidParams(
            "resize_within_bounding_box".into(),
            "width is required".into(),
        )
    })?;
    let max_h = p.height.ok_or_else(|| {
        HandlerError::InvalidParams(
            "resize_within_bounding_box".into(),
            "height is required".into(),
        )
    })?;
    let img = load(&bytes, mime, "resize_within_bounding_box")?;
    // `resize` (vs `resize_exact`) preserves the aspect ratio — the
    // canonical "fit within bounding box" semantic.
    let resized = img.resize(max_w, max_h, parse_filter(p.filter.as_deref()));
    let format = format_for_mime(mime, "resize_within_bounding_box")?;
    Ok(HandlerOutput {
        output_mime_type: mime.to_string(),
        output_bytes: Some(encode(resized, format)?),
        output_json: None,
    })
}

#[derive(Debug, Deserialize, Default)]
struct RotateParams {
    /// Degrees, clockwise. Only quarter-turns are exact-pixel
    /// operations; arbitrary angles fall back to a Lanczos resample
    /// of a 90-rotated image.
    degrees: Option<i32>,
}

pub fn rotate(mime: &str, params: &serde_json::Value, bytes: Vec<u8>) -> HandlerResult {
    let p: RotateParams = serde_json::from_value(params.clone())
        .map_err(|err| HandlerError::InvalidParams("rotate".into(), err.to_string()))?;
    let img = load(&bytes, mime, "rotate")?;
    let rotated = match p.degrees.unwrap_or(0).rem_euclid(360) {
        0 => img,
        90 => img.rotate90(),
        180 => img.rotate180(),
        270 => img.rotate270(),
        other => {
            return Err(HandlerError::InvalidParams(
                "rotate".into(),
                format!("only 0/90/180/270 deg supported, got {other}"),
            ));
        }
    };
    let format = format_for_mime(mime, "rotate")?;
    Ok(HandlerOutput {
        output_mime_type: mime.to_string(),
        output_bytes: Some(encode(rotated, format)?),
        output_json: None,
    })
}

#[derive(Debug, Deserialize, Default)]
struct CropParams {
    x: u32,
    y: u32,
    width: u32,
    height: u32,
}

pub fn crop(mime: &str, params: &serde_json::Value, bytes: Vec<u8>) -> HandlerResult {
    let p: CropParams = serde_json::from_value(params.clone())
        .map_err(|err| HandlerError::InvalidParams("crop".into(), err.to_string()))?;
    let img = load(&bytes, mime, "crop")?;
    if p.x.checked_add(p.width).map(|v| v > img.width()).unwrap_or(true)
        || p.y
            .checked_add(p.height)
            .map(|v| v > img.height())
            .unwrap_or(true)
    {
        return Err(HandlerError::InvalidParams(
            "crop".into(),
            "crop rectangle exceeds image bounds".into(),
        ));
    }
    let cropped = img.crop_imm(p.x, p.y, p.width, p.height);
    let format = format_for_mime(mime, "crop")?;
    Ok(HandlerOutput {
        output_mime_type: mime.to_string(),
        output_bytes: Some(encode(cropped, format)?),
        output_json: None,
    })
}

pub fn grayscale(mime: &str, bytes: Vec<u8>) -> HandlerResult {
    let img = load(&bytes, mime, "grayscale")?;
    let gray = img.grayscale();
    let format = format_for_mime(mime, "grayscale")?;
    Ok(HandlerOutput {
        output_mime_type: mime.to_string(),
        output_bytes: Some(encode(gray, format)?),
        output_json: None,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    /// 1×1 white PNG (smallest legal PNG with the right magic + IHDR).
    /// The image crate decodes / re-encodes it without losing the
    /// pixel, so resize / rotate / crop assertions stay deterministic.
    fn one_pixel_png() -> Vec<u8> {
        // Pre-computed: 1×1 RGBA white. Generated once with `image`
        // and pasted here so the test is hermetic.
        let img =
            image::DynamicImage::ImageRgba8(image::RgbaImage::from_pixel(1, 1, image::Rgba([255, 255, 255, 255])));
        encode(img, ImageFormat::Png).unwrap()
    }

    #[test]
    fn thumbnail_stays_inside_max_dim() {
        let bytes = encode(
            DynamicImage::ImageRgba8(image::RgbaImage::from_pixel(
                512,
                256,
                image::Rgba([0, 0, 0, 255]),
            )),
            ImageFormat::Png,
        )
        .unwrap();
        let out = thumbnail("image/png", &serde_json::json!({"max_dim": 64}), bytes).unwrap();
        let decoded = image::load_from_memory(out.output_bytes.as_ref().unwrap()).unwrap();
        assert!(decoded.width() <= 64 && decoded.height() <= 64);
    }

    #[test]
    fn rotate_90_swaps_axes() {
        let bytes = encode(
            DynamicImage::ImageRgba8(image::RgbaImage::from_pixel(
                4,
                2,
                image::Rgba([0, 0, 0, 255]),
            )),
            ImageFormat::Png,
        )
        .unwrap();
        let out = rotate("image/png", &serde_json::json!({"degrees": 90}), bytes).unwrap();
        let decoded = image::load_from_memory(out.output_bytes.as_ref().unwrap()).unwrap();
        assert_eq!((decoded.width(), decoded.height()), (2, 4));
    }

    #[test]
    fn grayscale_preserves_dimensions() {
        let bytes = one_pixel_png();
        let out = grayscale("image/png", bytes).unwrap();
        let decoded = image::load_from_memory(out.output_bytes.as_ref().unwrap()).unwrap();
        assert_eq!((decoded.width(), decoded.height()), (1, 1));
    }
}

File: openfoundry-go/services/media-transform-runtime-service/internal/handlers/image_ops.go
// Package handlers hosts the native (pure-Go) image transformations
// + the dispatch table the REST runtime uses to route a `kind` to its
// implementation. External-binary handlers (`ffmpeg`, `tesseract`, …)
// are intentionally absent; the runtime returns 501 for those kinds
// via the catalog without reaching this package.
package handlers

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"image"
	"image/gif"
	"image/jpeg"
	"image/png"

	"github.com/HugoSmits86/nativewebp"
	"github.com/disintegration/imaging"
	"golang.org/x/image/bmp"
	"golang.org/x/image/tiff"
	"golang.org/x/image/webp"
)

// HandlerOutput is what a handler hands back to the REST layer.
type HandlerOutput struct {
	OutputMimeType string
	OutputBytes    []byte
	OutputJSON     any
}

// Sentinel error categories. Wrap with %w (or use the helpers) so the
// runtime can map to the matching wire-format code.
var (
	ErrUnsupportedMime = errors.New("unsupported mime type")
	ErrInvalidParams   = errors.New("invalid params")
	ErrDecode          = errors.New("decode failed")
	ErrEncode          = errors.New("encode failed")
)

// imageFormat mirrors the discriminator the Rust runtime uses to drive
// the codec (`image::ImageFormat`). The MIME → format map matches the
// Rust `format_for_mime` exactly.
type imageFormat int

const (
	formatPNG imageFormat = iota
	formatJPEG
	formatWebP
	formatGIF
	formatTIFF
	formatBMP
)

func formatForMime(mime, kind string) (imageFormat, error) {
	switch mime {
	case "image/png":
		return formatPNG, nil
	case "image/jpeg", "image/jpg":
		return formatJPEG, nil
	case "image/webp":
		return formatWebP, nil
	case "image/gif":
		return formatGIF, nil
	case "image/tiff":
		return formatTIFF, nil
	case "image/bmp":
		return formatBMP, nil
	default:
		return 0, fmt.Errorf("%w `%s` for transformation `%s`", ErrUnsupportedMime, mime, kind)
	}
}

func decodeImage(src []byte, mime, kind string) (image.Image, imageFormat, error) {
	f, err := formatForMime(mime, kind)
	if err != nil {
		return nil, 0, err
	}
	r := bytes.NewReader(src)
	var img image.Image
	switch f {
	case formatPNG:
		img, err = png.Decode(r)
	case formatJPEG:
		img, err = jpeg.Decode(r)
	case formatGIF:
		img, err = gif.Decode(r)
	case formatWebP:
		img, err = webp.Decode(r)
	case formatTIFF:
		img, err = tiff.Decode(r)
	case formatBMP:
		img, err = bmp.Decode(r)
	}
	if err != nil {
		return nil, 0, fmt.Errorf("%w: %s", ErrDecode, err.Error())
	}
	return img, f, nil
}

func encodeImage(img image.Image, f imageFormat) ([]byte, error) {
	var buf bytes.Buffer
	var err error
	switch f {
	case formatPNG:
		err = png.Encode(&buf, img)
	case formatJPEG:
		err = jpeg.Encode(&buf, img, nil)
	case formatGIF:
		err = gif.Encode(&buf, img, nil)
	case formatWebP:
		// Pure-Go WebP encoder (lossless VP8L). Matches Rust output
		// shape — a WebP-in / WebP-out round trip stays WebP.
		err = nativewebp.Encode(&buf, img, nil)
	case formatTIFF:
		err = tiff.Encode(&buf, img, nil)
	case formatBMP:
		err = bmp.Encode(&buf, img)
	}
	if err != nil {
		return nil, fmt.Errorf("%w: %s", ErrEncode, err.Error())
	}
	return buf.Bytes(), nil
}

// resizeParams mirrors the Rust struct of the same name.
type resizeParams struct {
	Width  *uint32 `json:"width,omitempty"`
	Height *uint32 `json:"height,omitempty"`
	// Default Lanczos3 — Foundry's "Resize" produces high-fidelity
	// downscales; matches the upstream sample renders.
	Filter *string `json:"filter,omitempty"`
}

func parseFilter(name *string) imaging.ResampleFilter {
	want := "lanczos3"
	if name != nil {
		want = *name
	}
	switch want {
	case "nearest":
		return imaging.NearestNeighbor
	case "triangle":
		// Rust `image::FilterType::Triangle` is the linear-tap filter.
		return imaging.Linear
	case "catmullrom":
		return imaging.CatmullRom
	case "gaussian":
		return imaging.Gaussian
	default:
		return imaging.Lanczos
	}
}

func invalidParams(kind, msg string) error {
	return fmt.Errorf("%w for `%s`: %s", ErrInvalidParams, kind, msg)
}

// Thumbnail returns a fixed-box thumbnail. Foundry's default longest
// edge is 256 px; `max_dim` overrides it.
func Thumbnail(mime string, params json.RawMessage, src []byte) (HandlerOutput, error) {
	img, f, err := decodeImage(src, mime, "thumbnail")
	if err != nil {
		return HandlerOutput{}, err
	}
	maxDim := uint32(256)
	if len(params) > 0 {
		var p struct {
			MaxDim *uint32 `json:"max_dim,omitempty"`
		}
		if err := json.Unmarshal(params, &p); err != nil {
			return HandlerOutput{}, invalidParams("thumbnail", err.Error())
		}
		if p.MaxDim != nil {
			maxDim = *p.MaxDim
		}
	}
	// imaging.Thumbnail crops to the exact box; for parity with Rust's
	// `DynamicImage::thumbnail` (which preserves aspect ratio inside
	// a max×max box) we use Fit instead.
	scaled := imaging.Fit(img, int(maxDim), int(maxDim), imaging.Lanczos)
	out, err := encodeImage(scaled, f)
	if err != nil {
		return HandlerOutput{}, err
	}
	return HandlerOutput{OutputMimeType: mime, OutputBytes: out}, nil
}

// Resize matches Rust's `resize_exact` — width and height are honoured
// independently (no aspect-ratio preservation).
func Resize(mime string, params json.RawMessage, src []byte) (HandlerOutput, error) {
	var p resizeParams
	if len(params) > 0 {
		if err := json.Unmarshal(params, &p); err != nil {
			return HandlerOutput{}, invalidParams("resize", err.Error())
		}
	}
	img, f, err := decodeImage(src, mime, "resize")
	if err != nil {
		return HandlerOutput{}, err
	}
	bounds := img.Bounds()
	tw := uint32(bounds.Dx())
	th := uint32(bounds.Dy())
	if p.Width != nil {
		tw = *p.Width
	}
	if p.Height != nil {
		th = *p.Height
	}
	resized := imaging.Resize(img, int(tw), int(th), parseFilter(p.Filter))
	out, err := encodeImage(resized, f)
	if err != nil {
		return HandlerOutput{}, err
	}
	return HandlerOutput{OutputMimeType: mime, OutputBytes: out}, nil
}

// ResizeWithinBoundingBox preserves aspect ratio within a max W×H box.
// width and height are required (Rust enforces the same).
func ResizeWithinBoundingBox(mime string, params json.RawMessage, src []byte) (HandlerOutput, error) {
	var p resizeParams
	if len(params) > 0 {
		if err := json.Unmarshal(params, &p); err != nil {
			return HandlerOutput{}, invalidParams("resize_within_bounding_box", err.Error())
		}
	}
	if p.Width == nil {
		return HandlerOutput{}, invalidParams("resize_within_bounding_box", "width is required")
	}
	if p.Height == nil {
		return HandlerOutput{}, invalidParams("resize_within_bounding_box", "height is required")
	}
	img, f, err := decodeImage(src, mime, "resize_within_bounding_box")
	if err != nil {
		return HandlerOutput{}, err
	}
	resized := imaging.Fit(img, int(*p.Width), int(*p.Height), parseFilter(p.Filter))
	out, err := encodeImage(resized, f)
	if err != nil {
		return HandlerOutput{}, err
	}
	return HandlerOutput{OutputMimeType: mime, OutputBytes: out}, nil
}

// Rotate honours only quarter-turns (0/90/180/270 deg, clockwise) —
// arbitrary angles are rejected with a 400, matching the Rust handler.
func Rotate(mime string, params json.RawMessage, src []byte) (HandlerOutput, error) {
	deg := 0
	if len(params) > 0 {
		var p struct {
			Degrees *int `json:"degrees,omitempty"`
		}
		if err := json.Unmarshal(params, &p); err != nil {
			return HandlerOutput{}, invalidParams("rotate", err.Error())
		}
		if p.Degrees != nil {
			deg = *p.Degrees
		}
	}
	// Mirror Rust's `rem_euclid(360)` so negative degrees normalise.
	deg = ((deg % 360) + 360) % 360

	img, f, err := decodeImage(src, mime, "rotate")
	if err != nil {
		return HandlerOutput{}, err
	}
	var rotated image.Image
	switch deg {
	case 0:
		rotated = img
	case 90:
		// Rust `image::DynamicImage::rotate90` rotates 90° clockwise;
		// disintegration/imaging's `Rotate90` rotates 90° counter-
		// clockwise. Use `Rotate270` to match Rust's clockwise semantic.
		rotated = imaging.Rotate270(img)
	case 180:
		rotated = imaging.Rotate180(img)
	case 270:
		rotated = imaging.Rotate90(img)
	default:
		return HandlerOutput{}, invalidParams("rotate", fmt.Sprintf("only 0/90/180/270 deg supported, got %d", deg))
	}
	out, err := encodeImage(rotated, f)
	if err != nil {
		return HandlerOutput{}, err
	}
	return HandlerOutput{OutputMimeType: mime, OutputBytes: out}, nil
}

// Crop returns the rectangle (x, y, width, height); bounds are
// validated so an out-of-range crop is a 400, matching Rust.
func Crop(mime string, params json.RawMessage, src []byte) (HandlerOutput, error) {
	var p struct {
		X      uint32 `json:"x"`
		Y      uint32 `json:"y"`
		Width  uint32 `json:"width"`
		Height uint32 `json:"height"`
	}
	if len(params) == 0 {
		return HandlerOutput{}, invalidParams("crop", "x, y, width, height required")
	}
	if err := json.Unmarshal(params, &p); err != nil {
		return HandlerOutput{}, invalidParams("crop", err.Error())
	}
	img, f, err := decodeImage(src, mime, "crop")
	if err != nil {
		return HandlerOutput{}, err
	}
	w := uint32(img.Bounds().Dx())
	h := uint32(img.Bounds().Dy())
	xEnd, okX := safeAdd(p.X, p.Width)
	yEnd, okY := safeAdd(p.Y, p.Height)
	if !okX || !okY || xEnd > w || yEnd > h {
		return HandlerOutput{}, invalidParams("crop", "crop rectangle exceeds image bounds")
	}
	rect := image.Rect(int(p.X), int(p.Y), int(xEnd), int(yEnd))
	cropped := imaging.Crop(img, rect)
	out, err := encodeImage(cropped, f)
	if err != nil {
		return HandlerOutput{}, err
	}
	return HandlerOutput{OutputMimeType: mime, OutputBytes: out}, nil
}

// Grayscale emits a luminance-only render of the input. No params.
func Grayscale(mime string, src []byte) (HandlerOutput, error) {
	img, f, err := decodeImage(src, mime, "grayscale")
	if err != nil {
		return HandlerOutput{}, err
	}
	gray := imaging.Grayscale(img)
	out, err := encodeImage(gray, f)
	if err != nil {
		return HandlerOutput{}, err
	}
	return HandlerOutput{OutputMimeType: mime, OutputBytes: out}, nil
}

func safeAdd(a, b uint32) (uint32, bool) {
	r := a + b
	if r < a {
		return 0, false
	}
	return r, true
}

// Dispatch routes a request to the matching native handler. Catalogue-
// only kinds (External / NotImplemented) never reach this function —
// the REST layer short-circuits them with the catalog status.
func Dispatch(kind, mime string, params json.RawMessage, src []byte) (HandlerOutput, error) {
	switch kind {
	case "thumbnail":
		return Thumbnail(mime, params, src)
	case "resize":
		return Resize(mime, params, src)
	case "resize_within_bounding_box":
		return ResizeWithinBoundingBox(mime, params, src)
	case "rotate":
		return Rotate(mime, params, src)
	case "crop":
		return Crop(mime, params, src)
	case "grayscale":
		return Grayscale(mime, src)
	case "embedding":
		return Embedding(mime, params, src)
	case "transcription":
		return Transcription(mime, params, src)
	case "layout_aware_v2":
		return LayoutAwareV2(mime, params, src)
	case "vlm_extract":
		return VLMExtract(mime, params, src)
	default:
		return HandlerOutput{}, invalidParams(kind, "no native handler — caller should not have reached Dispatch()")
	}
}
